src/lib.rs

```diff
@@ -108,27 +106,48 @@ pub enum FoundCrate {
 /// The returned crate name is sanitized in such a way that it is a valid rust identifier. Thus,
 /// it is ready to be used in `extern crate` as identifier.
 pub fn crate_name(orig_name: &str) -> Result<FoundCrate, Error> {
-    let manifest_dir =
-        PathBuf::from(env::var("CARGO_MANIFEST_DIR").map_err(|_| Error::CargoManifestDirNotSet)?);
+    struct Cache {
+        manifest_dir: String,
+        manifest_path: PathBuf,
+        manifest: Table,
```

Comment:
Before we merge this, could you try to directly cache all crate names. So, have a hash table here that maps from crate name to crate name in the cargo file? By all I mean all crates that are referenced in the cargo.toml. May also brings some improvements?

Reply:
I implemented this. I think there is actually a behavioral change though: When you have multiple dependencies on the same package, previously the first one in the order they were searched was used to determine the crate name, now it's the last one in the order they are iterated to build the package > crate mapping (because, AFAIK, when `collect()`ing a `BTreeMap` the last occurrence of any given key "wins").

I'm pretty sure those two orderings are the same the way I implemented it now, so when a package is depended on with different crate names, a different one will be returned from `crate_name` now. I think assume this acceptable, just wanted to spell it out.

Reply:
Yeah that sounds right, but this is also some very niche use case that wasn't tested before.

Reply:
BTW, did that improve the performance?

Reply:
I can benchmark again next week.

Reply:
Would be nice

Reply:
This change doesn't seem to have a measureable impact on performance.